language-plutus-core/prelude/PlutusPrelude.hs

```diff
@@ -226,6 +227,11 @@ foldMapM :: (Foldable f, Monad m, Monoid b) => (a -> m b) -> f a -> m b
 foldMapM f xs = foldr step return xs mempty where
     step x r z = f x >>= \y -> r $! z `mappend` y
 
+-- | This function generalizes 'eitherToMaybe', 'eitherToList',
+-- 'listToMaybe' and other such functions.
+reoption :: (Foldable f, Alternative g) => f a -> g a
```

Comment:
`Control.Applicative.Alternative.afold`?

Reply:
Wasn't aware of that. It takes more indirections, though:

```
afromList :: Alternative f => [a] -> f a
afromList = Foldable.asum . map pure

afold :: (Alternative f, Foldable t) => t a -> f a
afold = afromList . Foldable.toList
```

I like the direct implementation more.

Reply:
I also like the direct implementation more, but I like using library functions more :)

Reply:
Okay, my trump card: I like the `reoption` name more, because it indicates what is happening: a transformation from one type treated as optional to another type treated the same way.

Note also that `afold` behaves differently: `asum` calls `(<|>)`, so in the end we have `pure x <|> r`, but this is only equal to `x` for alternatives that obey the left catch law and many collections do not obey it, so for example using `afold` over a tree is going to return all the elements of a tree in a, say, list. While `reoption` would return the head of the tree as a singleton list.

I think both operation are sensible and useful, but right now I'm using only the option types mapping behavior, so I prefer to keep `reoption`.